File: 76-minimum-window-substring/minimum-window-substring.java
class Solution {
    public String minWindow(String s, String t) {
        // Edge case: if s is shorter than t, no valid window possible
        if (s.length() < t.length()) return "";

        // Step 1: count how many of each character we NEED (from t)
        HashMap<Character, Integer> need = new HashMap<>();
        for (int i = 0; i < t.length(); i++) {
            char c = t.charAt(i);
            need.put(c, need.getOrDefault(c, 0) + 1);
        }

        // Step 2: tracks counts of characters currently in our window
        HashMap<Character, Integer> window = new HashMap<>();

        int required = need.size();   // unique chars we need to fully match
        int formed = 0;               // unique chars currently fully matched

        int left = 0;
        int minLen = Integer.MAX_VALUE;  // smallest valid window length found
        int minStart = 0;                // starting index of smallest window

        // Step 3: expand window using 'right'
        for (int right = 0; right < s.length(); right++) {
            char c = s.charAt(right);
            window.put(c, window.getOrDefault(c, 0) + 1);

            // if this character's count now matches what's needed, one more requirement satisfied
            if (need.containsKey(c) && window.get(c).intValue() == need.get(c).intValue()) {
                formed++;
            }

            // Step 4: window is valid (contains all of t) -> try shrinking from left
            while (formed == required) {
                // update answer if current window is smaller than best found so far
                if (right - left + 1 < minLen) {
                    minLen = right - left + 1;
                    minStart = left;
                }

                // remove leftmost character from window
                char leftChar = s.charAt(left);
                window.put(leftChar, window.get(leftChar) - 1);

                // if removing it breaks a requirement, window becomes invalid
                if (need.containsKey(leftChar) && window.get(leftChar) < need.get(leftChar)) {
                    formed--;
                }

                left++;  // shrink window
            }
        }

        // Step 5: return smallest valid window found, or "" if none
        return minLen == Integer.MAX_VALUE ? "" : s.substring(minStart, minStart + minLen);
    }
}

File: 76-minimum-window-substring/minimum-window-substring.py
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        if len(s) < len(t):
            return ""

        window_count = {}
        have = 0
        left = 0
        min_len = float('inf')
        min_left = 0

        t_count = {}
        for ch in t:
            if ch in t_count:
                t_count[ch] += 1
            else:
                t_count[ch] = 1

        need = len(t_count)

        for right in range(len(s)):
            char = s[right]
            if char in window_count:
                window_count[char] += 1
            else:
                window_count[char] = 1

            if char in t_count and window_count[char] == t_count[char]:
                have += 1

            while have == need:
                current_len = right - left + 1
                if current_len < min_len:
                    min_len = current_len
                    min_left = left

                left_char = s[left]
                window_count[left_char] -= 1
                if left_char in t_count and window_count[left_char] < t_count[left_char]:
                    have -= 1
                left += 1

        if min_len == float('inf'):
            return ""
        return s[min_left : min_left + min_len]